sdk/core/System.ClientModel/src/Message/PipelineRequestHeaders.cs

```diff
@@ -0,0 +1,140 @@
+        {
+            values = GetHeaderValueEnumerable(name, value);
+            return true;
+        }
+
+        values = default;
+        return false;
+    }
+
+    public override IEnumerator<KeyValuePair<string, string>> GetEnumerator()
+        => GetHeadersStringValues().GetEnumerator();
+
+    // Internal API provided to take advantage of performance-optimized implementation.
+    internal bool GetNextValue(int index, out string name, out object value)
+    {
+        if (index >= _headers.Count)
+        {
+            name = default!;
+            value = default!;
+            return false;
+        }
+
+        _headers.GetAt(index, out IgnoreCaseString headerName, out object headerValue);
+        name = headerName;
+        value = headerValue;
+        return true;
+    }
+
+    #region Implementation
+    private IEnumerable<KeyValuePair<string, string>> GetHeadersStringValues()
+    {
+        for (int i = 0; i < _headers.Count; i++)
+        {
+            _headers.GetAt(i, out IgnoreCaseString name, out object value);
+            string values = GetHeaderValueString(name, value);
+            yield return new KeyValuePair<string, string>(name, values);
+        }
+    }
+
+    private IEnumerable<KeyValuePair<string, IEnumerable<string>>> GetHeadersListValues()
+    {
+        for (int i = 0; i < _headers.Count; i++)
+        {
+            _headers.GetAt(i, out IgnoreCaseString name, out object value);
+            IEnumerable<string> values = GetHeaderValueEnumerable(name, value);
+            yield return new KeyValuePair<string, IEnumerable<string>>(name, values);
+        }
+    }
+
+    private static string GetHeaderValueString(string name, object value)
+        => value switch
+        {
+            string stringValue => stringValue,
+            List<string> listValue => string.Join(",", listValue),
+            _ => throw new InvalidOperationException($"Unexpected type for header {name}: {value?.GetType()}")
+        };
+
+    private static IEnumerable<string> GetHeaderValueEnumerable(string name, object value)
+        => value switch
+        {
+            string stringValue => new[] { stringValue },
+            List<string> listValue => listValue,
+            _ => throw new InvalidOperationException($"Unexpected type for header {name}: {value.GetType()}")
+        };
+
+    private readonly struct IgnoreCaseString : IEquatable<IgnoreCaseString>
+    {
+        private readonly string _value;
+
+        public IgnoreCaseString(string value)
+        {
+            _value = value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public bool Equals(IgnoreCaseString other) => string.Equals(_value, other._value, StringComparison.OrdinalIgnoreCase);
+        public override bool Equals(object? obj) => obj is IgnoreCaseString other && Equals(other);
+        public override int GetHashCode() => _value.GetHashCode();
```

Comment:
could it cause a problem that this is case sensitive?

Reply:
Interesting!  I think this is the same as what we have in Azure.Core today - let me look at it a little.

Reply:
yeah, it's a bug. We probably never use gethashcode and so the bug never manifested as a problem.

Reply:
maybe to future proof ourselves, we should throw here or do something like:
```c#
public override int GetHashCode() => _value switch
{
    string s => s.ToLower().GetHashCode(),
    _ => throw new UnsupportedOperationException("...")
};
```

Reply:
Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41232